session/pingpong/invoice_tracker.go

```diff
@@ -154,57 +137,153 @@ type InvoiceTrackerDeps struct {
-		timeTracker:                    itd.TimeTracker,
-		paymentInfo:                    itd.PaymentInfo,
-		providerID:                     itd.ProviderID,
-		accountantCaller:               itd.AccountantCaller,
-		accountantPromiseStorage:       itd.AccountantPromiseStorage,
-		accountantID:                   itd.AccountantID,
+		deps:                           itd,
 		maxNotReceivedExchangeMessages: calculateMaxNotReceivedExchangeMessageCount(chargePeriodLeeway, itd.ChargePeriod),
-		maxAccountantFailureCount:      itd.MaxAccountantFailureCount,
-		maxAllowedAccountantFee:        itd.MaxAllowedAccountantFee,
-		bcHelper:                       itd.BlockchainHelper,
-		publisher:                      itd.Publisher,
-		registryAddress:                itd.Registry,
-		feeProvider:                    itd.FeeProvider,
-		channelAddressCalculator:       itd.ChannelAddressCalculator,
-		maxRRecoveryLength:             itd.MaxRRecoveryLength,
+		invoicesSent:                   make(map[string]sentInvoice),
 	}
 }
 
 func calculateMaxNotReceivedExchangeMessageCount(chargeLeeway, chargePeriod time.Duration) uint64 {
 	return uint64(math.Round(float64(chargeLeeway) / float64(chargePeriod)))
 }
 
-func (it *InvoiceTracker) generateInitialInvoice() error {
-	agreementID, err := it.invoiceStorage.GetNewAgreementID(it.providerID)
+func (it *InvoiceTracker) markInvoiceSent(invoice sentInvoice) {
+	it.invoiceLock.Lock()
+	defer it.invoiceLock.Unlock()
+
+	it.invoicesSent[invoice.invoice.Hashlock] = invoice
+}
+
+func (it *InvoiceTracker) markInvoicePaid(hashlock []byte) {
+	it.invoiceLock.Lock()
+	defer it.invoiceLock.Unlock()
+
+	delete(it.invoicesSent, hex.EncodeToString(hashlock))
+}
+
+func (it *InvoiceTracker) getMarkedInvoice(hashlock []byte) (invoice sentInvoice, ok bool) {
+	it.invoiceLock.Lock()
+	defer it.invoiceLock.Unlock()
+	in, ok := it.invoicesSent[hex.EncodeToString(hashlock)]
+	return in, ok
+}
+
+func (it *InvoiceTracker) listenForExchangeMessages() error {
+	for {
+		select {
+		case pm := <-it.deps.ExchangeMessageChan:
+			err := it.handleExchangeMessage(pm)
+			if err != nil && err != ErrInvoiceExpired {
+				return err
+			}
+		case <-it.stop:
+			return nil
+		}
+	}
+}
+
+func (it *InvoiceTracker) handleExchangeMessage(pm crypto.ExchangeMessage) error {
+	invoice, ok := it.getMarkedInvoice(pm.Promise.Hashlock)
+	if !ok {
+		log.Debug().Msgf("consumer sent exchange message with missing expired hashlock %v, skipping", invoice.invoice.Hashlock)
+		return ErrInvoiceExpired
+	}
+
+	err := it.validateExchangeMessage(pm)
 	if err != nil {
-		return errors.Wrap(err, "could not get new agreement id")
+		return err
 	}
 
-	r := it.generateR()
-	invoice := crypto.CreateInvoice(agreementID, it.paymentInfo.GetPrice().Amount, 0, r)
-	invoice.Provider = it.providerID.Address
-	it.lastInvoice = lastInvoice{
-		invoice: invoice,
-		r:       r,
+	it.lastExchangeMessage = pm
+	it.markInvoicePaid(pm.Promise.Hashlock)
```

Comment:
Do `markInvoicePaid` need mutex if handleExchangeMessage is executed in single goroutine?

Reply:
there are 3 routines - one is sending the invoices and marking them as sent, while the other is receiving the exchange messages and marking the relevant invoices as sent. There's also a timeout which is running on a separate routine as well.